Dev libfabric

stephenmsachs opened the pull request:
- Add `+external-libfabric` variant to `intel-oneapi-mpi` & `intel-mpi` packages:

If `+external-libfabric` is set, a libfabric.so from the external `libfabric`
package is added to RUNPATH instead of Intel MPIs internal one.

- Align `intel-mpi` with `intel-oneapi-mpi` package behavior:

Intel MPI used to add its libfabric path to 'SPACK_COMPILER_EXTRA_RPATHS' and
therefore the RPATH of every dependent built with `intel-mpi`. Intel OneAPI MPI
added its libfabric path to the RUNPATH of libmpi.so.

Commit:
[Tutorials] Update elpa checksums to current working ones

The elpa checksums have changed on https://elpa.mpcdf.mpg.de/
Also, the original download location
http://elpa.mpcdf.mpg.de/elpa-2015.11.001.tar.gz will produce a broken tar file.

Commit:
[Tutorials] Update elpa checksums to current working ones

The elpa checksums have changed on https://elpa.mpcdf.mpg.de/
Also, the original download location
http://elpa.mpcdf.mpg.de/elpa-2015.11.001.tar.gz will produce a broken tar file.

Commit:
Merge branch 'develop' of https://github.com/spack/spack into develop

Commit:
[Tutorials] Update Armadillo checksum to current working ones

`spack fetch armadillo` fails with "all fetchers failed" before this patch.

Commit:
[Tutorials] Update Armadillo checksum to current working ones

`spack fetch armadillo` fails with "all fetchers failed" before this patch.

Commit:
Merge branch 'develop' of https://github.com/spack/spack into develop

Commit:
Merge branch 'develop' of https://github.com/spack/spack into develop

Commit:
Merge branch 'develop' of https://github.com/spack/spack into develop

Commit:
[intel-oneapi-mpi] Obey I_MPI_OFI_LIBRARY_INTERNAL setting

- Do not put internal libfabric as RPATH into libmpi.so
- Only load internal path to libfabric if I_MPI_OFI_LIBRARY_INTERNAL not set to 0.

Commit:
[intel-oneapi-mpi] Obey I_MPI_OFI_LIBRARY_INTERNAL setting

- Do not put internal libfabric as RPATH into libmpi.so
- Only load internal path to libfabric if I_MPI_OFI_LIBRARY_INTERNAL not set to 0.

Commit:
[intel-oneapi-mpi] Fix typo

Commit:
[intel-oneapi-mpi] Fix typo

Commit:
[intel-oneapi-mpi] Fix: default for environment variable not set

Commit:
[intel-oneapi-mpi] Fix: default for environment variable not set

Commit:
[intel-oneapi-mpi] Never add libfabric to RPATH

Commit:
[intel-oneapi-mpi] Never add libfabric to RPATH

Commit:
[intel-oneapi-mpi] Revert removing libfabric rpath

patchelf puts RUNPATH into libmpi.so which can be overwritten by LD_LIBRARY_PATH

Commit:
[intel-oneapi-mpi] Revert removing libfabric rpath

patchelf puts RUNPATH into libmpi.so which can be overwritten by LD_LIBRARY_PATH

Commit:
[intel-oneapi-mpi] Add +libfabric for external libfabric dependency

Commit:
[intel-oneapi-mpi] Add +libfabric for external libfabric dependency

Commit:
[intel-oneapi-mpi] Change variant identifier to '+external-libfabric'

This mimics the behavior of other optional external dependencies

Commit:
[intel-oneapi-mpi] Change variant identifier to '+external-libfabric'

This mimics the behavior of other optional external dependencies

Commit:
Add '+external-libfabric' to intel-mpi and align to intel-oneapi-mpi

- Add +external-libfabric variant to intel-mpi

If '+external-libfabric' is set, the libfabric.so from the external libfabric
package is added to RUNPATH instead of Intel MPIs internal one.

- Align intel-mpi with intel-oneapi-behavior:

Intel MPI used to add its libfabric path to 'SPACK_COMPILER_EXTRA_RPATHS' and
therefore the RPATH of every dependent built with intel-mpi. Intel OneAPI MPI
added its libfabric path to the RUNPATH of libmpi.so.

cr: https://code.amazon.com/reviews/CR-59149206

Commit:
Add '+external-libfabric' to intel-mpi and align to intel-oneapi-mpi

- Add +external-libfabric variant to intel-mpi

If '+external-libfabric' is set, the libfabric.so from the external libfabric
package is added to RUNPATH instead of Intel MPIs internal one.

- Align intel-mpi with intel-oneapi-behavior:

Intel MPI used to add its libfabric path to 'SPACK_COMPILER_EXTRA_RPATHS' and
therefore the RPATH of every dependent built with intel-mpi. Intel OneAPI MPI
added its libfabric path to the RUNPATH of libmpi.so.

cr: https://code.amazon.com/reviews/CR-59149206

Commit:
Add '+external-libfabric' to intel-mpi and align to intel-oneapi-mpi

- Add +external-libfabric variant to intel-mpi

If '+external-libfabric' is set, the libfabric.so from the external libfabric
package is added to RUNPATH instead of Intel MPIs internal one.

- Align intel-mpi with intel-oneapi-behavior:

Intel MPI used to add its libfabric path to 'SPACK_COMPILER_EXTRA_RPATHS' and
therefore the RPATH of every dependent built with intel-mpi. Intel OneAPI MPI
added its libfabric path to the RUNPATH of libmpi.so.

Commit:
Merge branch 'dev-libfabric' of ssh://git.amazon.com/pkg/InfosphereSpack into dev-libfabric

Commit:
Revert "[Tutorials] Update Armadillo checksum to current working ones"

This reverts commit 8d36c5d9f1929cec4c8a39658441dcd51c9f3406.

Commit:
Revert "[Tutorials] Update elpa checksums to current working ones"

This reverts commit e951f2d7eba635c6954a05bbdd108ec2c07a00e8.

Commit:
Merge branch 'dev-libfabric' of ssh://git.amazon.com/pkg/InfosphereSpack into dev-libfabric

Commit:
Revert "[Tutorials] Update elpa checksums to current working ones"

This reverts commit e951f2d7eba635c6954a05bbdd108ec2c07a00e8.

Commit:
Revert "[Tutorials] Update Armadillo checksum to current working ones"

This reverts commit 8d36c5d9f1929cec4c8a39658441dcd51c9f3406.

Commit:
Make 'linux tests / style' pass

rscohn2 commented on var/spack/repos/builtin/packages/intel-mpi/package.py:
```diff
@@ -75,3 +82,21 @@ def setup_run_environment(self, env):
+                )
+            for libmpi in glob.glob(join_path(self.component_lib_dir('mpi'),
+                                              '**', 'libmpi*.so')):
+                subprocess.call(['patchelf', '--set-rpath', libfabric_rpath, libmpi])
```
Does it need a build dependence on patchelf in this file?

stephenmsachs replied:
Good catch, thank you. I'll add that.

rscohn2 commented on lib/spack/spack/build_systems/intel.py:
```diff
@@ -1091,15 +1091,6 @@ def _setup_dependent_env_callback(
-                # Intel MPI since 2019 depends on libfabric which is not in the
-                # lib directory but in a directory of its own which should be
-                # included in the rpath
-                if self.version_yearlike >= ver('2019'):
```
It would be good to get an opinion from @scheibelp on using patchelf vs SPACK_COMPILER_EXTRA_RPATHS. I did the patchelf because I was already using a similar mechanism in the compiler package.

scheibelp replied:
> It would be good to get an opinion from @scheibelp on using patchelf vs SPACK_COMPILER_EXTRA_RPATHS.

Generally I don't recommend manipulating `SPACK_COMPILER_EXTRA_RPATHS` directly: it's an "internal" variable (i.e. we could change how we handle it in such a way that edits like this might not have any effect).

Example alternative: you could get a similar effect by adding a `libs` method which returns this directory: Spack will automatically RPATH all dependency library directories (one issue with this approach is that all Packages inheriting from  `IntelPackage` would have to be careful to replicate or call the `super()` of `libs`. Note this is different to 

> I did the patchelf because I was already using a similar mechanism in the compiler package.

since in this case we are in `setup_dependent_env` configuring the build for dependents, and I think in the case where patchelf was used, we were modifying a build directly (i.e. not as a dependent).

stephenmsachs replied:
One more difference I found is that `patchelf --set-rpath` actually sets the `DT_RUNPATH` inside `libmpi.so`. So the location of libfabric could be overridden by `LD_LIBRARY_PATH`. Libraries added to the `libs()` method are, if I am not mistaken, burned into the `RPATH` of the dependent package's executables.

rscohn2 replied:
I have some other pending changes for oneapi libraries: #26256 that I need to finish up and could implement Peter's recommendation there. I am still busy with other work so I cannot do it right away--probably a week or so. Or if you want to make the change in this PR, that would be great.

stephenmsachs replied:
I have removed the `patchelf` calls in intel-mpi and intel-oneapi-mpi and replaced it with additions to `libs()`. I have made the changes as far up the call stack as I think is reasonable, i.e. in `build_systems/intel.py` and `intel-oneapi-mpi/package.py`.

My checks with gromacs and osu microbenchmarks picked up the correct libfabric.

rscohn2 reviewed (dismissed):
Thanks for the contribution.

Commit:
Make patchelf available at `intel-mpi` package build time

Commit:
Move from using `patchelf` to `libs()` for libfabric

This change will no longer patch the libmpi.so `RUNPATH` with libfabric.so, but
instead add libfabric.so to the `RPATH` of intel{-oneapi}-mpi dependent packages.

Commit:
Fix typo

rscohn2 commented on var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py:
```diff
@@ -87,17 +88,20 @@ def libs(self):
+            libfabric_rpath = os.path.dirname(
+                [s for s in self.spec['libfabric'].libs if s.endswith('libfabric.so')][0]
+            )
+        libs += libfabric_rpath
```
Why do you only take libfabric.so and not the full self.spec['libfabric'].libs?

stephenmsachs replied:
I agree, if that place exists it would make a better candidate. I have not been able to find a libs() function (or property) in the inheritance chain Libfabric<AutoToolsPackage<PackageBase. 

@rajachan Do you think it is worth adding libs() function to the libfabric package?

rajachan replied:
Sure, go for it. If the oneapi packages are hitting this issue, others might too so it will be good to enumerate any custom paths via libs().

That said, all libs should already get installed inside the libfabric prefix's `lib` dir by default. 

```
$ tree install 
install
|-- bin
|   |-- fi_info
|   |-- fi_pingpong
|   `-- fi_strerror
|-- include
|   `-- rdma
|       |-- fabric.h
|       |-- fi_atomic.h
|       |-- fi_cm.h
|       |-- fi_collective.h
|       |-- fi_domain.h
|       |-- fi_endpoint.h
|       |-- fi_eq.h
|       |-- fi_errno.h
|       |-- fi_ext.h
|       |-- fi_rma.h
|       |-- fi_tagged.h
|       `-- fi_trigger.h
|-- lib
|   |-- libfabric.a
|   |-- libfabric.la
|   |-- libfabric.so -> libfabric.so.1.17.0
|   |-- libfabric.so.1 -> libfabric.so.1.17.0
|   |-- libfabric.so.1.17.0
|   `-- pkgconfig
|       `-- libfabric.pc

<snip>
```

I say that without reading the changes in the PR though, will take a closer look at this later today.

stephenmsachs replied:
So you are saying libs found in `/installation/root/lib` are added to the `libs()` function by default? That would make sense.

scheibelp replied:
> So you are saying libs found in /installation/root/lib are added to the libs() function by default? That would make sense.

If you're asking about `libfabric`, then yes: Because the library name `libfabric.so` matches the name of the package (`libfabric`), this will be handled automatically.

stephenmsachs replied:
OK, so I am using the full `self.spec['libfabric'].libs` now. No changes to the libfabric package necessary.

rscohn2 reviewed (commented):
This looks good to me. Just one question in the review.

Commit:
Add full libfabric libraries to RPATH for `+external-libfabric`

rscohn2 commented on lib/spack/spack/build_systems/intel.py:
```diff
@@ -994,6 +994,16 @@ def libs(self):
+            if self.version_yearlike >= ver('2019'):
+                # Patch libmpi.so rpath so it can find libfabric
+                d = ancestor(self.component_lib_dir('mpi'))
+                libfabric_rpath = os.path.join(d, 'libfabric', 'lib')
```
```suggestion
                libfabric_rpath = [os.path.join(d, 'libfabric', 'lib')]
```
later code assumes this is a list

stephenmsachs committed:
Add libfabric path as list to `libs()`

Co-authored-by: Robert Cohn <rscohn2@gmail.com>

stephenmsachs commented:
@rscohn2 are we closing this one for https://github.com/spack/spack/pull/27292 ?

rscohn2 commented:
If you don't mind, I would like to do that. I was running ```spack style``` to fix the style errors and it was reporting that many files were modified, even though git status did not show it. Then I tried rebasing and it messed up all the changes. I started with a clean checkout of develop and manually moved the changes over and that corrected the strange behavior with spack style.

I changed the logic for adding libfabric because there were some issues when I tested for using the internal libfabric.

stephenmsachs commented:
Of course, thanks for taking the lead. I'll keep improving my git fluency so that the next PR will go smoother.
